rust/kernel/error.rs

```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
+where
+    T: TryFrom<c_types::c_int>,
+    T::Error: core::fmt::Debug,
+{
+    match r {
+        Ok(v) => v,
+        Err(e) => T::try_from(e.to_kernel_errno()).unwrap(),
+    }
+}
+
+/// Transforms a `crate::error::Result<T>` to a kernel `C` integer result.
+///
+/// This is useful when calling Rust functions that return `Result<T>`
+/// from inside `extern "C"` functions that need to return an integer
+/// error result.
+///
+/// `T` should be convertible to an integer via `TryFrom<c_types::c_int>`.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// unsafe extern "C" fn probe_callback(
+///     pdev: *mut bindings::platform_device,
+/// ) -> c_types::c_int {
+///     from_kernel_result! {
+///         let ptr = devm_alloc(pdev)?;
+///         rust_helper_platform_set_drvdata(pdev, ptr);
+///         Ok(0)
+///     }
+/// }
+/// ```
```

Comment:
In general, please split commits if you are doing several changes. e.g. the documentation addition (which is great!) should be on its own commit.